File: ios/Shared/WssTicketClient.swift
import Foundation

struct WssSessionTicket: Equatable, Sendable, CustomStringConvertible, CustomDebugStringConvertible {
    /// Opaque bearer credential. It must never enter a URL, log, metric or error message.
    let ticket: String
    let expiresAt: Date
    /// Broker-echoed URL, compared byte-for-byte with the selected signed front by the caller.
    let url: String

    func isFresh(at now: Date) -> Bool {
        expiresAt > now
    }

    var description: String {
        "WssSessionTicket(ticket: <redacted>, expiresAt: \(expiresAt), url: <redacted>)"
    }

    var debugDescription: String { description }
}

struct WssTicketStatusError: Error, Equatable {
    let status: Int
    let retryAfterMilliseconds: UInt64?
}

struct WssTicketPolicy: Equatable, Sendable {
    let totalDeadlineMilliseconds: UInt64
    let perAttemptMilliseconds: UInt64
    let defaultRetryAfterMilliseconds: UInt64
    let maxRetryAfterMilliseconds: UInt64

    init(
        totalDeadlineMilliseconds: UInt64 = 15_000,
        perAttemptMilliseconds: UInt64 = 5_000,
        defaultRetryAfterMilliseconds: UInt64 = 10_000,
        maxRetryAfterMilliseconds: UInt64 = 30_000
    ) {
        precondition(totalDeadlineMilliseconds > 0)
        precondition(perAttemptMilliseconds > 0)
        precondition(defaultRetryAfterMilliseconds > 0)
        precondition(maxRetryAfterMilliseconds > 0)
        self.totalDeadlineMilliseconds = totalDeadlineMilliseconds
        self.perAttemptMilliseconds = perAttemptMilliseconds
        self.defaultRetryAfterMilliseconds = defaultRetryAfterMilliseconds
        self.maxRetryAfterMilliseconds = maxRetryAfterMilliseconds
    }
}

typealias WssTicketAttempt = @Sendable (
    _ brokerURL: URL,
    _ relayID: String,
    _ frontID: String,
    _ clientID: String?,
    _ sessionID: String?,
    _ timeoutMilliseconds: UInt64
) async throws -> WssSessionTicket

/// Policy-preserving WSS ticket client. Front ordering, total/per-attempt budgets, first-error
/// semantics, and the bounded 429/503 retry round remain Swift-owned; each individual transport
/// attempt is one fresh brokerapi operation.
final class WssTicketClient: @unchecked Sendable {
    private let operationFactory: any NativeBrokerOperationFactory

    init(operationFactory: any NativeBrokerOperationFactory) {
        self.operationFactory = operationFactory
    }

    func requestWithFailover(
        brokerURLs: [URL],
        relayID: String,
        frontID: String,
        clientID: String? = nil,
        sessionID: String? = nil
    ) async throws -> WssSessionTicket {
        try await Self.requestWithFailover(
            brokerURLs: brokerURLs,
            relayID: relayID,
            frontID: frontID,
            clientID: clientID,
            sessionID: sessionID,
            policy: WssTicketPolicy(),
            monotonicMilliseconds: {
                UInt64(max(ProcessInfo.processInfo.systemUptime * 1_000, 0))
            },
            wait: { milliseconds in
                let nanoseconds = milliseconds.multipliedReportingOverflow(by: 1_000_000)
                try await Task.sleep(
                    nanoseconds: nanoseconds.overflow ? UInt64.max : nanoseconds.partialValue
                )
            },
            attempt: { [operationFactory] brokerURL, requestedRelayID, requestedFrontID,
                requestedClientID, requestedSessionID, timeout in
                try await Self.requestOnce(
                    operationFactory: operationFactory,
                    brokerURL: brokerURL,
                    relayID: requestedRelayID,
                    frontID: requestedFrontID,
                    clientID: requestedClientID,
                    sessionID: requestedSessionID,
                    timeoutMilliseconds: timeout
                )
            }
        )
    }

    /// Injectable orchestration core. Fronts are sequential under one deadline; only 429/503 may
    /// schedule one additional round, after a bounded Retry-After wait.
    static func requestWithFailover(
        brokerURLs: [URL],
        relayID: String,
        frontID: String,
        clientID: String?,
        sessionID: String?,
        policy: WssTicketPolicy,
        monotonicMilliseconds: @Sendable () -> UInt64,
        wait: @Sendable (UInt64) async throws -> Void,
        attempt: @escaping WssTicketAttempt
    ) async throws -> WssSessionTicket {
        guard relayID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false else {
            throw BrokerNativeFailure(kind: .validation, message: "The relay ID is empty.")
        }
        guard frontID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false else {
            throw BrokerNativeFailure(kind: .validation, message: "The WSS front ID is empty.")
        }
        var fronts: [URL] = []
        for url in brokerURLs where fronts.contains(url) == false {
            fronts.append(url)
        }
        guard fronts.isEmpty == false else {
            throw BrokerNativeFailure(kind: .validation, message: "No WSS ticket broker is available.")
        }

        let started = monotonicMilliseconds()
        let deadline = saturatedAdd(started, policy.totalDeadlineMilliseconds)
        var firstFailure: Error?
        var scheduledRetryDelay: UInt64?

        for round in 0...1 {
            for brokerURL in fronts {
                try Task.checkCancellation()
                let remaining = remainingMilliseconds(deadline: deadline, now: monotonicMilliseconds())
                guard remaining > 0 else { throw firstFailure ?? URLError(.timedOut) }
                let attemptTimeout = min(policy.perAttemptMilliseconds, remaining)
                do {
                    return try await withTimeout(milliseconds: attemptTimeout) {
                        try await attempt(
                            brokerURL,
                            relayID,
                            frontID,
                            clientID,
                            sessionID,
                            attemptTimeout
                        )
                    }
                } catch is CancellationError {
                    throw CancellationError()
                } catch let failure as BrokerNativeFailure where failure.isLocalPlatformFailure {
                    // Another front cannot repair a stale binding, nil result, or local schema
                    // mismatch. Abort before another short-lived bearer credential is minted.
                    throw failure
                } catch {
                    if firstFailure == nil { firstFailure = error }
                    if round == 0, let candidate = retryDelay(for: error, policy: policy) {
                        scheduledRetryDelay = max(scheduledRetryDelay ?? 0, candidate)
                    }
                }
            }

            if round == 1 { break }
            guard let delay = scheduledRetryDelay else {
                throw firstFailure ?? URLError(.cannotConnectToHost)
            }
            guard let firstFailure else { throw URLError(.cannotConnectToHost) }
            let remaining = remainingMilliseconds(deadline: deadline, now: monotonicMilliseconds())
            guard delay < remaining else { throw firstFailure }
            try await wait(delay)
            try Task.checkCancellation()
            guard remainingMilliseconds(deadline: deadline, now: monotonicMilliseconds()) > 0 else {
                throw firstFailure
            }
        }
        throw firstFailure ?? URLError(.cannotConnectToHost)
    }

    static func requestOnce(
        operationFactory: any NativeBrokerOperationFactory,
        brokerURL: URL,
        relayID: String,
        frontID: String,
        clientID: String? = nil,
        sessionID: String? = nil,
        timeoutMilliseconds _: UInt64 = 5_000
    ) async throws -> WssSessionTicket {
        let result: NativeBrokerWSSTicketResultSnapshot = try await NativeBrokerRunner.run(
            factory: operationFactory
        ) { operation in
            operation.requestWSSTicket(
                brokerURL: brokerURL.absoluteString,
                relayID: relayID,
                frontID: frontID,
                clientID: clientID ?? "",
                sessionID: sessionID ?? ""
            )
        }

        do {
            try result.throwIfFailed()
        } catch let failure as BrokerNativeFailure {
            if failure.isLocalPlatformFailure {
                throw failure
            }
            if let status = failure.httpStatus {
                throw WssTicketStatusError(
                    status: status,
                    retryAfterMilliseconds: failure.retryAfterMilliseconds
                )
            }
            throw failure
        }

        // Foundation's Date accepts non-finite/out-of-domain values. Keep the conversion bounded
        // to positive Unix milliseconds through the end of year 9999.
        let maximumUnixMilliseconds: Int64 = 253_402_300_799_999
        guard (1...maximumUnixMilliseconds).contains(result.expiresAtMilliseconds) else {
            throw BrokerNativeFailure(
                kind: .decode,
                message: "The native WSS ticket expiry is invalid."
            )
        }
        let seconds = Double(result.expiresAtMilliseconds) / 1_000
        guard seconds.isFinite else {
            throw BrokerNativeFailure(
                kind: .decode,
                message: "The native WSS ticket expiry is invalid."
            )
        }
        let expiresAt = Date(timeIntervalSince1970: seconds)
        guard expiresAt.timeIntervalSince1970.isFinite else {
            throw BrokerNativeFailure(
                kind: .decode,
                message: "The native WSS ticket expiry is invalid."
            )
        }
        return WssSessionTicket(ticket: result.ticket, expiresAt: expiresAt, url: result.url)
    }

    private static func retryDelay(for error: Error, policy: WssTicketPolicy) -> UInt64? {
        guard let status = error as? WssTicketStatusError, status.status == 429 || status.status == 503 else {
            return nil
        }
        let requested = status.retryAfterMilliseconds.flatMap { $0 > 0 ? $0 : nil }
            ?? policy.defaultRetryAfterMilliseconds
        return min(requested, policy.maxRetryAfterMilliseconds)
    }

    private static func withTimeout<T: Sendable>(
        milliseconds: UInt64,
        operation: @escaping @Sendable () async throws -> T
    ) async throws -> T {
        try await withThrowingTaskGroup(of: T.self) { group in
            group.addTask { try await operation() }
            group.addTask {
                let nanoseconds = milliseconds.multipliedReportingOverflow(by: 1_000_000)
                try await Task.sleep(
                    nanoseconds: nanoseconds.overflow ? UInt64.max : nanoseconds.partialValue
                )
                throw URLError(.timedOut)
            }
            defer { group.cancelAll() }
            guard let first = try await group.next() else { throw URLError(.timedOut) }
            return first
        }
    }

    private static func saturatedAdd(_ value: UInt64, _ increment: UInt64) -> UInt64 {
        value.addingReportingOverflow(increment).overflow ? UInt64.max : value + increment
    }

    private static func remainingMilliseconds(deadline: UInt64, now: UInt64) -> UInt64 {
        now >= deadline ? 0 : deadline - now
    }
}

File: android/app/src/main/java/com/openrung/net/WssTicketClient.kt
package com.openrung.net

import android.os.SystemClock
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withTimeout
import java.io.IOException
import java.net.SocketTimeoutException
import java.time.DateTimeException
import java.time.Instant
import kotlin.math.min

/** One short-lived credential bound by the broker to an exact relay and signed WSS front. */
data class WssSessionTicket(
    /** Opaque bearer value. It must never be logged or placed in a URL. */
    val ticket: String,
    val expiresAt: Instant,
    /** Exact front URL echoed by the broker; the caller compares it byte-for-byte. */
    val url: String,
) {
    /** Redact both sensitive values from data-class logging and debugger string surfaces. */
    override fun toString(): String =
        "WssSessionTicket(ticket=<redacted>, expiresAt=$expiresAt, url=<redacted>)"
}

/**
 * Typed status projection used by the existing bounded retry policy. Native error detail is kept
 * only as a bounded cause and neither the ticket nor bound URL is included in this message.
 */
class WssTicketStatusException(
    val status: Int,
    val retryAfterMillis: Long?,
    cause: Throwable? = null,
) : IOException("broker WSS ticket request failed with HTTP status $status", cause)

/** Bounds one complete broker-front failover attempt, including its optional retry round. */
internal data class WssTicketPolicy(
    val totalDeadlineMillis: Long = 15_000,
    val perAttemptMillis: Long = 5_000,
    val defaultRetryAfterMillis: Long = 10_000,
    val maxRetryAfterMillis: Long = 30_000,
) {
    init {
        require(totalDeadlineMillis > 0) { "WSS ticket total deadline must be positive" }
        require(perAttemptMillis > 0) { "WSS ticket attempt deadline must be positive" }
        require(defaultRetryAfterMillis > 0) { "WSS ticket default retry delay must be positive" }
        require(maxRetryAfterMillis > 0) { "WSS ticket maximum retry delay must be positive" }
    }
}

internal typealias WssTicketAttempt = suspend (
    brokerUrl: String,
    relayId: String,
    frontId: String,
    clientId: String?,
    sessionId: String?,
    timeoutMillis: Long,
) -> WssSessionTicket

/**
 * WSS-ticket policy stays in Kotlin; only each individual request is delegated to brokerapi.
 *
 * Front ordering, one total budget, one optional retry round, first-error semantics, and 429/503
 * eligibility are intentionally unchanged.
 */
object WssTicketClient {
    suspend fun requestWithFailover(
        brokerUrls: List<String>,
        relayId: String,
        frontId: String,
        clientId: String? = null,
        sessionId: String? = null,
    ): WssSessionTicket = requestWithFailover(
        brokerUrls = brokerUrls,
        relayId = relayId,
        frontId = frontId,
        clientId = clientId,
        sessionId = sessionId,
        policy = WssTicketPolicy(),
        elapsedRealtimeMillis = SystemClock::elapsedRealtime,
        wait = { delay(it) },
        attempt = { brokerUrl, requestedRelayId, requestedFrontId, requestedClientId, requestedSessionId, timeout ->
            requestOnce(
                brokerUrl = brokerUrl,
                relayId = requestedRelayId,
                frontId = requestedFrontId,
                clientId = requestedClientId,
                sessionId = requestedSessionId,
                timeoutMillis = timeout,
            )
        },
    )

    /** Injectable pure orchestration core retained for hostless policy tests. */
    internal suspend fun requestWithFailover(
        brokerUrls: List<String>,
        relayId: String,
        frontId: String,
        clientId: String?,
        sessionId: String?,
        policy: WssTicketPolicy,
        elapsedRealtimeMillis: () -> Long,
        wait: suspend (Long) -> Unit,
        attempt: WssTicketAttempt,
    ): WssSessionTicket {
        require(relayId.isNotBlank()) { "WSS ticket relay_id is required" }
        require(frontId.isNotBlank()) { "WSS ticket front_id is required" }
        val fronts = brokerUrls.map(String::trim).filter(String::isNotEmpty).distinct()
        require(fronts.isNotEmpty()) { "no broker fronts configured for WSS ticket" }

        val startedAt = elapsedRealtimeMillis()
        val deadline = saturatedAdd(startedAt, policy.totalDeadlineMillis)
        var firstFailure: Throwable? = null
        var retryDelayMillis: Long? = null

        for (round in 0..1) {
            for (brokerUrl in fronts) {
                currentCoroutineContext().ensureActive()
                val remainingMillis = remainingMillis(deadline, elapsedRealtimeMillis())
                if (remainingMillis <= 0) throw firstFailure ?: ticketDeadlineExceeded()
                val attemptMillis = min(policy.perAttemptMillis, remainingMillis)

                val failure = try {
                    return withTimeout(attemptMillis) {
                        attempt(
                            brokerUrl,
                            relayId,
                            frontId,
                            clientId,
                            sessionId,
                            attemptMillis,
                        )
                    }
                } catch (_: TimeoutCancellationException) {
                    SocketTimeoutException("WSS ticket broker attempt timed out")
                } catch (error: CancellationException) {
                    // Caller cancellation is exceptional and must never mint a ticket on a later
                    // front. The native runner has already dispatched Close before this propagates.
                    throw error
                } catch (error: BrokerNativeFailure) {
                    // These are Android artifact/model/configuration failures, not evidence that
                    // another remote front is reachable. Abort without consuming another ticket.
                    if (error.isLocalPlatformFailure) throw error
                    error
                } catch (error: LinkageError) {
                    throw unavailableNativeBrokerFailure(error)
                } catch (error: Throwable) {
                    if (error.hasLocalNativeBrokerFailure()) throw error
                    error
                }

                if (firstFailure == null) firstFailure = failure
                if (round == 0) {
                    retryDelayFor(failure, policy)?.let { candidateDelay ->
                        retryDelayMillis = maxOf(retryDelayMillis ?: 0, candidateDelay)
                    }
                }
            }

            if (round == 1) break
            val retryDelay = retryDelayMillis ?: throw checkNotNull(firstFailure)
            val remainingMillis = remainingMillis(deadline, elapsedRealtimeMillis())
            // A wait that consumes the whole remaining budget cannot leave time for another request.
            if (retryDelay >= remainingMillis) throw checkNotNull(firstFailure)
            wait(retryDelay)
            currentCoroutineContext().ensureActive()
            if (remainingMillis(deadline, elapsedRealtimeMillis()) <= 0) {
                throw checkNotNull(firstFailure)
            }
        }

        throw firstFailure ?: ticketDeadlineExceeded()
    }

    /**
     * Performs one synchronous native RequestWSSTicket under the caller's surrounding timeout.
     * Every invocation creates and closes a fresh operation.
     */
    internal suspend fun requestOnce(
        brokerUrl: String,
        relayId: String,
        frontId: String,
        clientId: String? = null,
        sessionId: String? = null,
        timeoutMillis: Long = WssTicketPolicy().perAttemptMillis,
        operationFactory: NativeBrokerOperationFactory = NativeBrokerTransport.operationFactory,
    ): WssSessionTicket {
        require(relayId.isNotBlank()) { "WSS ticket relay_id is required" }
        require(frontId.isNotBlank()) { "WSS ticket front_id is required" }
        require(timeoutMillis > 0) { "WSS ticket timeout must be positive" }

        val native = runNativeBrokerOperation(operationFactory) { operation ->
            operation.requestWSSTicket(
                brokerUrl = brokerUrl,
                relayId = relayId,
                frontId = frontId,
                clientId = clientId.orEmpty(),
                sessionId = sessionId.orEmpty(),
            )
        }
        val result = try {
            requireNativeBrokerSuccess(native, "native WSS ticket request")
        } catch (failure: BrokerNativeFailure) {
            if (failure.isLocalPlatformFailure) throw failure
            val status = failure.httpStatus
            if (status != null) {
                throw WssTicketStatusException(
                    status = status,
                    retryAfterMillis = failure.retryAfterMillis,
                    cause = failure,
                )
            }
            throw failure
        }

        val expiresAt = try {
            Instant.ofEpochMilli(result.expiresAtMillis)
        } catch (_: DateTimeException) {
            throw BrokerNativeFailure(
                kind = BrokerNativeFailureKind.DECODE,
                message = "native WSS ticket result has an invalid expiry",
            )
        }
        return WssSessionTicket(
            ticket = result.ticket,
            expiresAt = expiresAt,
            url = result.url,
        )
    }

    private fun retryDelayFor(error: Throwable, policy: WssTicketPolicy): Long? {
        val statusError = error as? WssTicketStatusException ?: return null
        if (statusError.status != 429 && statusError.status != 503) return null
        // Zero/negative native values were projected to null; absent hints use the existing delay.
        val requested = statusError.retryAfterMillis?.takeIf { it > 0 }
            ?: policy.defaultRetryAfterMillis
        return requested.coerceAtMost(policy.maxRetryAfterMillis)
    }

    private fun saturatedAdd(value: Long, increment: Long): Long =
        if (value > Long.MAX_VALUE - increment) Long.MAX_VALUE else value + increment

    private fun remainingMillis(deadline: Long, now: Long): Long =
        if (now >= deadline) 0 else deadline - now

    private fun ticketDeadlineExceeded(): SocketTimeoutException =
        SocketTimeoutException("WSS ticket request deadline exceeded")

    private fun Throwable.hasLocalNativeBrokerFailure(): Boolean {
        val seen = HashSet<Throwable>()
        var current: Throwable? = this
        while (current != null && seen.add(current)) {
            if (current is BrokerNativeFailure && current.isLocalPlatformFailure) return true
            current = current.cause
        }
        return false
    }
}
